### ISSUE TITLE
Selenium 4 - Node taking 3 minutes to be registered as UP

### ISSUE THREAD

**Issue by gtaujeky** [SeleniumHQ/selenium#9847]
## 🐛 Bug Report

Hi there,

I'm running Selenium Grid 4 using the fully distributed mode on K8s running on GKE.

However, I have some issue with node taking a long time to be marked as UP.

As Dynamic Grid (https://github.com/SeleniumHQ/docker-selenium#dynamic-grid-) is not currently supported in Kubernetes, I wrote a cronjob that will automatically scale up & down the number of node depending on the number of sessions running & sessions in the queue.

While the script is working fine (it's scaling up & down depending on the number of tests), our tests are taking a very long time to run as some nodes would take a long time to be ready.

## To Reproduce

I'm running the fully distributed mode using the following yaml file - https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/k8s-deployment-full-grid.yaml

Which is the same as the provided yaml file: https://github.com/SeleniumHQ/docker-selenium/blob/trunk/k8s-deployment-full-grid.yaml, I'm just adding namespace to make it easier to manage on my side.

I managed to reproduce this issue by manually scaling up/down the number of node (using `kubectl scale deployment`). 

In order to track this issue, I wrote a small bash script to see how many of nodes are registered and how many are up or down.
https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/check.sh

You can see all logs here
https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/check-30.log

Please note, I started from a fresh new deployment of selenium to my k8s cluster using `kubectl apply -f k8s-deployment-full-grid.yaml`

Events (extract from https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/check-30.log)

- `Mon Sep 20 11:41:25 AEST 2021`: Started deployment of Selenium Grid to K8s. No node is ready yet.
- `Mon Sep 20 11:41:48 AEST 2021`: Node is ready and marked as UP
- `Mon Sep 20 11:44:51 AEST 2021`: Scaled up the number of replicas to 30 (`kubectl scale --replicas=30 deployment`)
- `Mon Sep 20 11:44:57 AEST 2021`: After a few seconds, all 30 are marked as `UP`
- `Mon Sep 20 11:47:04 AEST 2021`: Scaled down the number of replicas to 1.
- `Mon Sep 20 11:50:54 AEST 2021`: Still waiting for grid to remove 2 node marked as `DOWN`
- `Mon Sep 20 11:50:59 AEST 2021`: Scaled back the number of replicas to 30.
- `Mon Sep 20 11:51:04 AEST 2021`: All 30 new nodes are added to the grid, but marked as `DOWN`
- `Mon Sep 20 11:53:53 AEST 2021`: Only about 3 minutes later, all 30 nodes are marked as `UP`

I captured logs from a few nodes that you can see here:
https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/chrome-node-1.log
https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/chrome-node-2.log
https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/chrome-node-3.log

As you can see, nodes have been added to the grid at
```
01:51:01.874 INFO [NodeServer.lambda$createHandlers$2] - Node has been added
```

But are only marked as `UP` about 3 minutes later.

You can find the full log of the distributor here: https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/distributor.log

I'm not sure if my issue is related to https://github.com/SeleniumHQ/docker-selenium/issues/1337, as I have this issue with the latest version (4.0.0-rc-2-prerelease-20210916).
I also tried to run the suggested version `4.0.0-rc-1-prerelease-20210618`, but have the same issue.

## Expected behavior

It should take that long for a node to be marked as `UP` and ready to run tests

## Test script or set of commands reproducing this issue

Please provide a test script to reproduce the issue you are reporting, if the 
setup is more complex, GitHub repo links with are also OK.

Please see above

## Environment

Selenium Grid version: 4.0.0-rc-2-prerelease-20210916

**Comment by diemol** [comment]
Usually fresh deployments need to pull the Docker images first so they can actually start the pods. What do you see in the pod logs? What is taking a long time?

**Comment by gtaujeky** [comment]
Hi @diemol,

Thanks for your reply.

Chrome node are up and running almost as soon as I scale up the number of pods

```
Mon Sep 20 11:50:59 AEST 2021: Scaled back the number of replicas to 30.
Mon Sep 20 11:51:04 AEST 2021: All 30 new nodes are added to the grid, but marked as DOWN
```

As you can see here https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/check-30.log#L449

I scaled up the number of node to be 30, and nodes have been added to the selenium grid.

https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/chrome-node-1.log#L125
https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/chrome-node-2.log#L125
https://github.com/gtaujeky/selenium-autoscaling-issue/blob/master/chrome-node-3.log#L125

So pods are up and running, but marked as DOWN (I'm getting the status of the nodes from the GraphQL endpoint).

I believe the issue is coming from the distributor. 

FYI, I did a few tests by always killing the distributor pod before doing the scale up of chrome nodes, and didn't have any issue by doing that. 

I can give another try with the latest version.

Let me know if you need more details, I'm happy to reproduce the issue to get more logs.

**Comment by daljitdokal** [comment]
We are running selenium 4.1.0 in openshift 4.6 and seeing the same 3 minutes delay on node chrome to become available. interestingly it is only on the 3rd and any subsequent attempt to deploy nodes to the hub. 
We use standard kubernetes deployments to create nodes as part of the pipeline.

If we do a fresh deployment of the hub, the first two deployments of the node pods come up as expected but from the 3rd deployment onwards nodes takes 3 minutes.

**Comment by csgbn** [comment]
Seeing the same thing, on the 3rd deployment it takes 3 minutes before the node  is marked UP from DOWN. We run hub on a separate vm from nodes, 4.1.0, does not matter if the hub is deployed distributed or not, nor does any change in settings for other timeouts like health check.

Below is a an example for what we see in the distributor log, extracted based on IP for such a node:

```
22:28:25.980 DEBUG [NettyConnectListener.onFailure] - Failed to recover from connect exception: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: /10.199.4.171:5555 with channel [id: 0xf9862e73]
22:28:25.980 DEBUG [AsyncCompletionHandler.onThrowable] - Connection refused: /10.199.4.171:5555
java.net.ConnectException: Connection refused: /10.199.4.171:5555
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: /10.199.4.171:5555
22:28:26.982 DEBUG [NettyConnectListener.writeRequest] - Using new Channel '[id: 0x3b8e8d49, L:/10.199.4.21:51662 - R:/10.199.4.171:5555]' for 'GET' to '/status'
host: 10.199.4.171:5555
22:28:27.098 DEBUG [ChannelManager.tryToOfferChannelToPool] - Adding key: http://10.199.4.171:5555 for channel [id: 0x3b8e8d49, L:/10.199.4.21:51662 - R:/10.199.4.171:5555]
22:28:27.099 DEBUG [GridModel.add] - Adding node with id 83d74dc3-d365-41e9-a147-25d6d33dc531 and URI http://10.199.4.171:5555
22:28:27.099 INFO [LocalDistributor.add] - Added node 83d74dc3-d365-41e9-a147-25d6d33dc531 at http://10.199.4.171:5555. Health check every 20s
22:28:28.164 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:29.264 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:30.364 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:31.464 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:32.567 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:33.664 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:34.764 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:35.864 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:36.964 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:38.064 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:39.164 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:40.265 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:41.364 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:42.464 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:43.564 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:44.664 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:45.764 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:46.864 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:47.964 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:49.071 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:50.164 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:51.264 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:52.364 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:53.464 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:54.564 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:55.664 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:56.764 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:57.864 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:28:58.964 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:00.064 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:01.163 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:02.264 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:03.364 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:04.464 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:05.564 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:06.664 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:07.764 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:08.864 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:09.964 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:11.064 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:12.164 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:13.264 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:14.364 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:15.464 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:16.564 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:17.664 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:18.764 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:19.864 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:20.964 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:22.064 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:23.164 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:24.264 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:25.364 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:26.464 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:27.564 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.4.171:5555 : 1
22:29:27.564 DEBUG [DefaultChannelPool$IdleChannelDetector.expiredChannels] - Adding Candidate expired Channel [id: 0x3b8e8d49, L:/10.199.4.21:51662 - R:/10.199.4.171:5555] isIdleTimeoutExpired=true isRemotelyClosed=false isTtlExpired=false
22:29:27.564 DEBUG [DefaultChannelPool$IdleChannelDetector.closeChannels] - Closing Idle Channel [id: 0x3b8e8d49, L:/10.199.4.21:51662 - R:/10.199.4.171:5555]
22:29:27.564 DEBUG [AsyncHttpClientHandler.channelInactive] - Channel Closed: [id: 0x3b8e8d49, L:/10.199.4.21:51662 ! R:/10.199.4.171:5555] with attribute DISCARD
22:31:20.407 DEBUG [NettyConnectListener.writeRequest] - Using new Channel '[id: 0x8707545a, L:/10.199.4.21:52400 - R:/10.199.4.171:5555]' for 'GET' to '/status'
host: 10.199.4.171:5555
22:31:20.417 DEBUG [ChannelManager.tryToOfferChannelToPool] - Adding key: http://10.199.4.171:5555 for channel [id: 0x8707545a, L:/10.199.4.21:52400 - R:/10.199.4.171:5555]
22:31:20.418 INFO [GridModel.setAvailability] - Switching node 83d74dc3-d365-41e9-a147-25d6d33dc531 (uri: http://10.199.4.171:5555) from DOWN to UP
```

**Comment by csgbn** [comment]
Same result with 4.1.1, linux and windows nodes. This is a blocker for us adopting v4 as we need the hub to run long term and add/remove nodes dynamically.

**Comment by diemol** [comment]
@gtaujeky @csgbn would it be possible for you to post updated logs? I suspect the problem happens because the initial healthcheck fails due to network issues in while the pod is coming up, and then things are ok when the next healthcheck happens (120 seconds/3 minutes later).

**Comment by csgbn** [comment]
I will see about getting fresh logs, however, the issue only happens on the third add of new nodes. We do not use docker/k8, just AWS vms with hub and nodes in separate vms in the same data center, no network issues ever observed. Each time a new node gets added it has a different IP and each node have 20 slots but only accepts 1 concurrent session at the time.

**Comment by diemol** [comment]
This is the connection failure I mean (from your log above), which causes the initial healthcheck to fail.
```
22:28:25.980 DEBUG [NettyConnectListener.onFailure] - Failed to recover from connect exception: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: /10.199.4.171:5555 with channel [id: 0xf9862e73]
22:28:25.980 DEBUG [AsyncCompletionHandler.onThrowable] - Connection refused: /10.199.4.171:5555
java.net.ConnectException: Connection refused: /10.199.4.171:5555
```

**Comment by csgbn** [comment]
Yes, and restarting the distributor by itself, resets the issue and now you get another 2 successful events of adding/removing nodes, and failure on the third event add new nodes...

**Comment by diemol** [comment]
I understand that. I only need more recent logs to match the assumptions I have regarding the initial healthcheck so I can figure out an improvement.

**Comment by csgbn** [comment]
Here are some fresh logs of a node being added when the issue is already present. This time it took more than 4 minutes to get to UP status and we see that the more add/remove events increase time to UP status.  Note that once the node actually registers it runs the test normally. Also, we can see previous nodes that has been removed show up in logs with connection failures, they just keep trying, never times out.

```
18:47:47.418 FINEST [AbstractInternalLogger.log] - [id: 0x120a7341] CONNECT: /10.199.2.123:5555
18:47:48.425 DEBUG [NettyConnectListener.writeRequest] - Using new Channel '[id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555]' for 'GET' to '/status'
18:47:48.425 FINEST [AbstractInternalLogger.log] - [id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555] WRITE: 102B
|00000040| 73 74 3a 20 31 30 2e 31 39 39 2e 32 2e 31 32 33 |st: 10.199.2.123|
18:47:48.425 FINEST [AbstractInternalLogger.log] - [id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555] FLUSH
18:47:48.425 FINEST [AbstractInternalLogger.log] - [id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555] ACTIVE
18:47:48.640 FINEST [AbstractInternalLogger.log] - [id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555] READ: 2048B
host: 10.199.2.123:5555
18:47:48.640 FINEST [AbstractInternalLogger.log] - [id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555] READ COMPLETE
18:47:48.640 FINEST [AbstractInternalLogger.log] - [id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555] READ: 11147B
18:47:48.640 DEBUG [ChannelManager.tryToOfferChannelToPool] - Adding key: http://10.199.2.123:5555 for channel [id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555]
18:47:48.641 FINEST [AbstractInternalLogger.log] - [id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555] READ COMPLETE
18:47:48.641 DEBUG [GridModel.add] - Adding node with id bd63cdad-9206-4b1c-a9c9-95d945c87b77 and URI http://10.199.2.123:5555
18:47:48.641 INFO [LocalDistributor.add] - Added node bd63cdad-9206-4b1c-a9c9-95d945c87b77 at http://10.199.2.123:5555. Health check every 120s
18:47:48.791 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:47:49.891 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:47:50.991 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:47:52.091 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:47:53.191 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:47:54.291 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:47:55.391 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:47:56.491 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:47:57.591 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:47:58.691 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:47:59.792 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:00.891 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:01.991 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:03.091 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:04.191 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:05.291 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:06.391 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:07.491 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:08.591 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:09.692 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:10.791 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:11.891 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:12.991 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:14.091 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:15.191 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:16.291 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:17.391 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:18.491 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:19.591 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:20.691 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:21.791 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:22.891 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:23.992 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:25.091 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:26.191 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:27.291 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:28.391 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:29.491 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:30.591 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:31.691 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:32.791 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:33.891 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:34.991 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:36.091 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:37.191 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:38.292 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:39.391 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:40.491 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:41.592 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:42.692 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:43.791 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:44.891 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:45.991 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:47.092 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:48.191 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:49.291 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:48:49.291 DEBUG [DefaultChannelPool$IdleChannelDetector.expiredChannels] - Adding Candidate expired Channel [id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555] isIdleTimeoutExpired=true isRemotelyClosed=false isTtlExpired=false
18:48:49.291 DEBUG [DefaultChannelPool$IdleChannelDetector.closeChannels] - Closing Idle Channel [id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555]
18:48:49.291 FINEST [AbstractInternalLogger.log] - [id: 0x120a7341, L:/10.199.4.21:62726 - R:/10.199.2.123:5555] CLOSE
18:48:49.291 FINEST [AbstractInternalLogger.log] - [id: 0x120a7341, L:/10.199.4.21:62726 ! R:/10.199.2.123:5555] INACTIVE
18:48:49.291 DEBUG [AsyncHttpClientHandler.channelInactive] - Channel Closed: [id: 0x120a7341, L:/10.199.4.21:62726 ! R:/10.199.2.123:5555] with attribute DISCARD
18:48:49.291 FINEST [AbstractInternalLogger.log] - [id: 0x120a7341, L:/10.199.4.21:62726 ! R:/10.199.2.123:5555] UNREGISTERED
18:53:02.921 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366] CONNECT: /10.199.2.123:5555
18:53:02.922 DEBUG [NettyConnectListener.writeRequest] - Using new Channel '[id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555]' for 'GET' to '/status'
18:53:02.922 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] WRITE: 102B
|00000040| 73 74 3a 20 31 30 2e 31 39 39 2e 32 2e 31 32 33 |st: 10.199.2.123|
18:53:02.922 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] FLUSH
18:53:02.922 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] ACTIVE
18:53:02.931 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] READ: 2048B
host: 10.199.2.123:5555
18:53:02.931 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] READ COMPLETE
18:53:02.931 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] READ: 11147B
18:53:02.932 DEBUG [ChannelManager.tryToOfferChannelToPool] - Adding key: http://10.199.2.123:5555 for channel [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555]
18:53:02.932 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] READ COMPLETE
18:53:02.932 INFO [GridModel.setAvailability] - Switching node bd63cdad-9206-4b1c-a9c9-95d945c87b77 (uri: http://10.199.2.123:5555) from DOWN to UP
18:53:03.392 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:04.491 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:05.369 DEBUG [NettyRequestSender.pollPooledChannel] - Using pooled Channel '[id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555]' for 'POST' to 'http://10.199.2.123:5555/se/grid/node/session'
18:53:05.369 DEBUG [NettyRequestSender.sendRequestWithOpenChannel] - Using open Channel [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] for POST '/se/grid/node/session'
18:53:05.370 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] WRITE: 170B
18:53:05.370 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] WRITE: 5B
18:53:05.370 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] WRITE: 364B
18:53:05.370 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] WRITE: 2B
18:53:05.370 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] FLUSH
18:53:05.370 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] WRITE: 5B
18:53:05.370 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] FLUSH
18:53:32.796 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] READ: 3545B
|00000db0| 30 32 66 31 30 2e 31 39 39 2e 32 2e 31 32 33 3a |02f10.199.2.123:|
host: 10.199.2.123:5555
18:53:32.796 DEBUG [ChannelManager.tryToOfferChannelToPool] - Adding key: http://10.199.2.123:5555 for channel [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555]
18:53:32.796 FINEST [AbstractInternalLogger.log] - [id: 0x7b2c2366, L:/10.199.4.21:63380 - R:/10.199.2.123:5555] READ COMPLETE
18:53:33.091 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:34.191 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:35.291 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:36.391 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:37.491 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:38.591 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:39.691 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:40.792 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:41.891 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:42.991 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:44.091 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:45.191 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:46.291 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:47.391 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:48.491 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:49.592 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:50.691 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:51.791 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:52.891 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:53.991 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:55.091 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:56.191 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:57.292 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:58.391 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:53:59.491 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:54:00.591 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:54:01.691 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
18:54:02.791 DEBUG [DefaultChannelPool$IdleChannelDetector.run] - Entry count for : http://10.199.2.123:5555 : 1
```

I hope this helps.

**Comment by diemol** [comment]
In this case I see it came up after 6 minutes, but no rejected connections in the log. Is it possible to have the whole log? Not sure if during the filtering some important details got lost.

**Comment by csgbn** [comment]
[distributor-gridlog.log](https://github.com/SeleniumHQ/selenium/files/7841769/distributor-gridlog.log)


Here is the complete log after a fresh grid start, adding/removing 6 nodes 5 times. Example IP 10.199.1.47 took more than 5 minutes to UP. Previously removed nodes keep appearing and failing connections. When a node vm is terminated we execute:

cURL --request DELETE 'http://localhost:5553/se/grid/distributor/node/<node-id>' --header 'X-REGISTRATION-SECRET: <secret>'

before the vm is terminated.

The hub is running distributed on latest java 11.

**Comment by csgbn** [comment]
Do let me know if you need any further logs or want me to try potential code fixes, thanks.

**Comment by diemol** [comment]
Thank you for the logs. Sadly, I did not get a lot of information that could confirm my assumption. 

Nevertheless, I strongly suspect the issue is that the initial healthcheck, which happens when the Node is being added, is failing because for whatever reason the Node could not be reached via HTTP. The thing is that we only run the healhcheck again until the configured interval has passed (typically 3 minutes).

I will rework how this is implemented and retry the healthcheck a few times during the Node registration, this should fix it.

**Comment by diemol** [comment]
I just made some changes that should help. Early next week, 4.1.2 will be released. I will leave this open to get feedback after the release.

**Comment by csgbn** [comment]
This particular issue seems to be have been solved with 4.1.2, unable to re-produce. However, while the nodes are now registered as "UP" they are not immediately put to work like v3 grid. Although this issue was also present with 4.1.1. You can have 100 nodes ready and registered to the hub and then run 100 tests in parallel, the queue is 0 but the test suite is only processed by 15-20 nodes, the rest of nodes never get any tests routed to them.

It is also apparent that the cpu usage on the hub is now at least double compared to 4.1.1 (and a lot more cpu usage while no tests are being executed but nodes registered) and compared to 4.1.1 we are seeing some freezes/timeouts now and then when nodes are added/removed while tests are being processed, but we are still trying to scope out these scenarios.

4x the hub cpu/memory resources does not have any impact on test concurrency.

**Comment by diemol** [comment]
Ok, so the main issue was fixed but now seems that we have another one, probably never reported before. 

Is there a way that you can report this in a new issue so we have enough data to check, please?

**Comment by csgbn** [comment]
Ok, opened #10332

**Comment by diemol** [comment]
Thank you.
